std/exception.d

```diff
@@ -2148,3 +2148,65 @@ pure nothrow @safe unittest
 
     auto infSlice = infinite[0 .. $]; // this would throw otherwise
 }
+
+
+/++
+    Convenience mixin for trivially sub-classing exceptions
+
+    Even trivially sub-classing an exception involves writing boilerplate code
+    for the constructor to: 1) correctly pass in the source file and line number
+    the exception was thrown from; 2) be usable with $(LREF enforce) which
+    expects exception constructors to take arguments in a fixed order. This
+    mixin provides that boilerplate code.
+
+    Note however that you need to mark the $(B mixin) line with at least a
+    minimal (i.e. just $(B ///) DDoc comment if you want the mixed-in
+    constructors to be documented in the newly created Exception subclass.
+ +/
+mixin template basicExceptionCtors()
+{
+    /++
+        Params:
+            msg  = The message for the exception.
+            file = The file where the exception occurred.
+            line = The line number where the exception occurred.
+            next = The previous exception in the chain of exceptions, if any.
+    +/
+    this(string msg, string file = __FILE__, size_t line = __LINE__,
+         Throwable next = null) @safe pure nothrow
+    {
+        super(msg, file, line, next);
+    }
+
+    /++
+        Params:
+            msg  = The message for the exception.
+            next = The previous exception in the chain of exceptions.
+            file = The file where the exception occurred.
+            line = The line number where the exception occurred.
+    +/
+    this(string msg, Throwable next, string file = __FILE__,
+         size_t line = __LINE__) @safe pure nothrow
+    {
+        super(msg, file, line, next);
+    }
+}
+
+///
+pure nothrow @safe unittest
+{
+    class MeaCulpa: Exception
+    {
+        ///
+        mixin basicExceptionCtors;
+    }
+
+    try
+        throw new MeaCulpa("test");
+    catch (Exception e) // Bugzilla 15467
+    {
+        assert(e.msg == "test");
```

Comment:
@jamadagni  Any function with no arguments can be called without parens, including functions which would normally take an argument but don't thanks to UFCS - e.g. `range.map!(a => a % 10)` instead of `range.map!(a => a % 10)()` or `map!(a => a % 10)(range)`. `@property` doesn't actually do much at this point. It was supposed to be the case that it would be required to call functions without parens (the ability to do so without `@property` predates `@property`) and that you couldn't call an `@property` function with parens, but that enforcement was never properly implemented, and once UFCS was introduced, folks didn't want to have to use parens for non-`@property` functions, since they thought that it was ugly when combined with the parens for the template arguments. There's some possibility that it will be enforced in the future that `@property` functions be called without parens or that the setter syntax require `@property`, but it's no longer expected that we'll ever require `@property` to be able to call a function without parens. And if it weren't for the fact that you can't have property functions which return callables (e.g.  delegate) without using `@property` and have it require that it be called without parens, then `@property` would probably be abandoned, because Andrei is not a big fan of it (and Walter probably isn't either, but I don't recall for sure - Andrei has been much more vocal about it).

But essentially, `@property` was introduced to fix the delegate problem and to move to a less ad-hoc implementation of property functions, but it was never finished. At this point, `@property` is primarily used to indicate how the API designer intended it to be used. It does affect a few things (like the type of a function), but it's mostly for show at this point. And it's not clear what the state of `@property` will ultimately be.